### PR TITLE
(BOLT-707) Add `apply` pre-docs

### PR DESCRIPTION
**Pull request by MikaelSmith**
Adds pre-docs for the `apply` feature. Note that the feature is not new
in the latest release, but has been incrementally developed over the
0.21 release series.

### CODE REVIEW

File: pre-docs/applying_manifest_blocks.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+- `_noop => true` will apply the manifest block in Puppet's noop mode, returning a report of the changes that would be made.
+- `_run_as => <user>` applies the manifest block as the specified user (on transports that support it).
+
+## Configuring Concurrency
+
+Each target requires a separate catalog be compiled with its unique facts and vars. Apply will compile and apply catalogs in parallel on the Bolt host. Concurrency of catalog compilation is controlled by a new `compile-concurrency` config option that's limited to 2x the number of threads your CPU can run concurrently, while catalog application uses Bolt's default thread pool controlled by the `concurrency` option.
+
+## Caveats
+
+Exported resources are not supported in apply blocks. Users should pass values directly instead of exporting and collecting them from PuppetDB.
```
Comment (adreyer) [inline pre-docs/applying_manifest_blocks.md]: Should we mention the puppetdb query function here or does that just confuse things?

If you need to interact with resources managed during a normal puppet run you can use puppetdb_query

Comment (MikaelSmith) [reply pre-docs/applying_manifest_blocks.md]: I'm not sure that's a common scenario, but I can include it.

File: pre-docs/applying_manifest_blocks.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
+# Applying Manifest Blocks
+
+Within a plan, you can use Bolt to apply blocks of Puppet code (manifest blocks) to remote nodes. Similar to the `puppet apply` command - that applies a standalone Puppet manifest to a local system - Bolt leverages manifest blocks to pass code to remote nodes from the command line. You can create manifest blocks that use existing content from the Forge, or mix declarative resource configuration via manifest blocks with procedural orchestration and action in a plan.
```
Comment (MikaelSmith) [inline pre-docs/applying_manifest_blocks.md]: @katelopresti I reworded this paragraph slightly to fit in more with how we introduce other concepts in the docs.

File: pre-docs/applying_manifest_blocks.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+## A short example
+
+To get started, we'll show creating a manifest for web server and running it as a plan in Bolt. We'll put this module in Bolt's default Boltdir (`~/.puppetlabs/bolt`).
+
+### nginx on Linux
+
+Setting up a web server using [nginx](https://nginx.org):
+
+1. Go to `~/.puppetlabs/bolt/modules`
+1. Create a new module: `pdk new module profiles` (or create `~/.puppetlabs/bolt/modules/profiles/plans`)
```
Comment (lucywyman) [inline pre-docs/applying_manifest_blocks.md]: This command doesn't actually make a `plans` dir, at least using the latest from https://puppet.com/download-puppet-development-kit

Comment (MikaelSmith) [reply pre-docs/applying_manifest_blocks.md]: That's fair. Do you think I need to call out creating the plans dir explicitly? I can drop `/plans` from this step, and the next says create a file in a non-existent directory.

Comment (lucywyman) [reply pre-docs/applying_manifest_blocks.md]: Mostly the "or create ~/.puppetlabs/bolt/modules/profiles/plans" made it seem like an exclusive or. I think "Create a new module and add a plans directory: `command` (or create `blee/blah`)" is slightly more clear :woman_shrugging: .

Comment (MikaelSmith) [reply pre-docs/applying_manifest_blocks.md]: Done.

File: pre-docs/applying_manifest_blocks.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+   ```
+1. Run `bolt plan run profiles::nginx_install --nodes <NODE NAME>`
+1. From a web browser, navigate to `<NODE NAME>` and you should see a page saying `hello!`
+
+### iis on Windows
+
+Setting up a web server using [iis](https://www.iis.net):
+
+1. Go to `~/.puppetlabs/bolt/modules`
+1. Create a new module: `pdk new module profiles` (or create `~/.puppetlabs/bolt/modules/profiles/plans`)
```
Comment (lucywyman) [inline pre-docs/applying_manifest_blocks.md]: Ugh, sorry to be so pedantic, but same line here

File: pre-docs/applying_manifest_blocks.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+It generates all the variables a `puppet apply` would include, so code can be reused between Bolt and Puppet. It then copies custom module content from Bolt's modulepath to the target and applies the catalog using Puppet.
+
+Almost all features of the Pupet language - classes, custom resource types, functions - can be used, with exceptions noted in [caveats](#caveats). The nginx example above could make use of the [puppet-nginx](https://github.com/voxpupuli/puppet-nginx) module for more complex deployments.
+
+When the catalog compiles and is executed successfully on all targets, apply returns the reports generated by applying the catalog on each node.
+
+## Using Hiera Data
+
+Hiera is a tool for looking up data based on properties of a particular target. It allows you to separate configuration from context-specific data, such as facts or name of a target.
```
Comment (lucywyman) [inline pre-docs/applying_manifest_blocks.md]: This is a great 2 sentence summation of hiera!

Comment (MikaelSmith) [reply pre-docs/applying_manifest_blocks.md]: I think the very end is a little fuzzy, going to switch `such as facts or name of a target` to `where-as context may be facts or name of a target`.

File: pre-docs/applying_manifest_blocks.md (hunk 2 above)
Comment (katelopresti) [reply pre-docs/applying_manifest_blocks.md]: Minor edit to the second sentence: 
Similar to the 'puppet apply' command, which applies a standalone Puppet manifest to a local system, the Bolt 'apply' command leverages manifest blocks to pass code to remote nodes from the command line.

File: pre-docs/applying_manifest_blocks.md [hunk 6]
```diff
@@ -1,6 +1,6 @@
 # Applying Manifest Blocks
 
-Within a plan, you can use Bolt to apply blocks of Puppet code (manifest blocks) to remote nodes. Similar to the `puppet apply` command, which applies a standalone Puppet manifest to a local system, the Bolt `apply` command leverages manifest blocks to pass code to remote nodes from the command line. You can create manifest blocks that use existing content from the Forge, or mix declarative resource configuration via manifest blocks with procedural orchestration and action in a plan.
+Within a plan, you can use Bolt to apply blocks of Puppet code (manifest blocks) to remote nodes. Similar to the `puppet apply` command, which applies a standalone Puppet manifest to a local system, the Bolt `apply` command leverages manifest blocks to pass code to remote nodes from the command line. You can create manifest blocks that use existing content from the Forge, or mix declarative resource configuration via manifest blocks with procedural orchestration and action in a plan. Most features of the Puppet language are availabe in a manifest block: classes, custom resource types, and functions. Exceptions are noted in [caveats](#caveats).
```
Comment (MikaelSmith) [inline pre-docs/applying_manifest_blocks.md]: Typo `availabe` -> `available`.

**Pull request merged**